pkg/webhook/mutating/pod.go

```diff
@@ -111,22 +111,30 @@ func PodCreateMutation(ctx context.Context, req *admission.Request, handler *Han
 		}
 	}
 
-	var networkType = ipamtypes.ParseNetworkTypeFromString(utils.PickFirstNonEmptyString(pod.Annotations[constants.AnnotationNetworkType], pod.Labels[constants.LabelNetworkType]))
+	var networkTypeFromPod = utils.PickFirstNonEmptyString(pod.Annotations[constants.AnnotationNetworkType], pod.Labels[constants.LabelNetworkType])
+	var networkType = ipamtypes.ParseNetworkTypeFromString(networkTypeFromPod)
 	if len(networkName) > 0 {
 		network := &networkingv1.Network{}
 		if err = handler.Client.Get(ctx, types.NamespacedName{Name: networkName}, network); err != nil {
 			return admission.Errored(http.StatusInternalServerError, err)
 		}
-		switch networkType {
-		case ipamtypes.Underlay:
-			klog.Infof("[mutating] patch pod %s/%s with selector of network %s", req.Namespace, req.Name, networkName)
-			pod = patchSelectorToPod(pod, network.Spec.NodeSelector)
-			mutated = true
-		case ipamtypes.Overlay:
-			// overlay network has a wide scheduling domain over the whole cluster
-			// no more selectors should be patched
-		default:
-			return admission.Errored(http.StatusBadRequest, fmt.Errorf("unknown network type %s", networkType))
+
+		if len(networkTypeFromPod) > 0 {
+			switch networkType {
+			case ipamtypes.Underlay:
+				klog.Infof("[mutating] patch pod %s/%s with selector of network %s", req.Namespace, req.Name, networkName)
+				pod = patchSelectorToPod(pod, network.Spec.NodeSelector)
+				mutated = true
+			case ipamtypes.Overlay:
+				// overlay network has a wide scheduling domain over the whole cluster
+				// no more selectors should be patched
+			default:
+				return admission.Errored(http.StatusBadRequest, fmt.Errorf("unknown network type %s", networkType))
+			}
+		} else {
+			// specified network takes higher priority than network type defaulting, if no network type specified
+			// from pod, then network type should inherit from network type of specified network from pod
+			networkType = ipamtypes.ParseNetworkTypeFromString(string(networkingv1.GetNetworkType(network)))
```

Comment:
If a pod is specified to an Underlay network while without specific network type, the steps of patching node selector seems to be lost here?

Reply:
good catch!